### ISSUE TITLE
Delayed Synapses (DDE support)

### ISSUE THREAD

**Issue by Fololo** [jaxleyverse/jaxley#618]
Hi @michaeldeistler and Jaxley team,

This might be slightly off-topic and a very general question, but I thought I'd dare ask:

Suppose I want to build a network of simple neurons with long-ranging projections, and I want to represent those long-ranging connections with "delayed" synaptic activations. In other words, I want to model the transmission time of the spike through the axon without detailed modeling of a cell's individual compartments, but by adding a delay property to all edges of the network. To simulate the network, I would now need to solve a delay-differential equation (DDE) that accounts for the delayed inputs.

Any hints on what parts of `Jaxley` I would have to modify to get this working? I suppose this would affect the `Module.step()` function and potentially other functions called within it, but would I also need to make changes to the `Synapse` and `Channel` interfaces, or do you think it could be done without altering them?

**Comment by michaeldeistler** [comment]
Hi Florian!

Indeed, delays are not supported yet, but it would be fantastic to have them! If you want to work on this, I suggest we have a zoom call to discuss how to best approach this (send me an email so we can schedule a meeting).

Looking forward!
Michael

**Comment by llandsmeer** [comment]
Hi both,

I have been benchmarking some spike event queue implementations in jax lately (including gradients through the delay and w.r.t. the delay), then just found this issue while going over the related works... The work is not fully done yet, but I guess it could be useful for this project (at least make sure we don't do double work 😄 )

Maybe it makes sense to meet up and/or collaborate?

Thanks,
Lennart

**Comment by michaeldeistler** [comment]
Hi Lennart!

this sounds great! I would be happy to collaborate on this, I just sent you an email (let me know if you did not receive it, then I probably had an old email...)

All the best
Michael

**Comment by Fololo** [comment]
Hi @llandsmeer and @michaeldeistler,

sound's great! I would be happy to join the meeting as well. Haven't had time to work on this yet, but would certainly be nice to get in touch :)

All the best, 
Florian

**Comment by llandsmeer** [comment]
Hi both,

Thanks for reaching out. I guess it's the EMC email? I'm not able to access it until next week, l.p.l.landsmeer[at]tudelft.nl should work still

Best,
Lennart

**Comment by michaeldeistler** [comment]
Yes it was that email. I now sent an email l.p.l.landsmeer[at]tudelft.nl, hope this works!

Michael

**Comment by llandsmeer** [comment]
@Fololo 

For the continuous signal case, this might be your solution:

```python
import os, sys
from jax import config
config.update("jax_enable_x64", True)
config.update("jax_platform_name", "cpu")

import typing
import matplotlib.pyplot as plt

import jax
import jax.numpy as jnp

import jaxley as jx
from jaxley.channels import Na, K, Leak
from jaxley.connect import fully_connect
from jaxley.synapses.synapse import Synapse

class ShiftRegister(Synapse):
    DUMMY = -10e10 # 'empty signal'
    def __init__(self, name: typing.Optional[str] = None, *, delay_steps: int):
        '''
        implements delay for continous signal
        delay_steps * dt = delay_ms
        '''
        super().__init__(name)
        prefix = self._name
        assert delay_steps >= 1
        self.delay_steps = delay_steps
        self.synapse_params = {
            f'{prefix}_weight': 0.01
        }
        self.synapse_states: dict[str, typing.Any] = {
            f'{prefix}_{i}': self.DUMMY for i in range(delay_steps)
        }

    def update_states(
        self,
        states: dict,
        delta_t: float,
        pre_voltage: float,
        post_voltage: float,
        params: dict,
    ) -> dict:
        prefix = self._name
        inp = pre_voltage
        return {
            f'{prefix}_{i}': states[f'{prefix}_{i+1}']
            if i < self.delay_steps - 1
            else inp
            for i in range(self.delay_steps)
        }

    def compute_current(
        self, states: dict, pre_voltage: float, post_voltage: float, params: dict
    ) -> float:
        prefix = self._name
        vprev_delayed = states[f'{prefix}_0']
        return jax.lax.select(vprev_delayed == self.DUMMY, # type: ignore
                - jnp.zeros_like(post_voltage),
                - params[f'{prefix}_weight'] * (vprev_delayed - post_voltage)
                              )


i_delay = 0.0   # ms
i_amp = 0.15    # nA
i_dur = 1.0     # ms
dt = 0.025      # ms
net = jx.Network([jx.Cell() for _ in range(3)])
fully_connect(net.cell(0), net.cell(1),
              ShiftRegister(delay_steps=int(10 / dt))) # 10 ms delay
fully_connect(net.cell(1), net.cell(2),
              ShiftRegister(delay_steps=int(10 / dt))) # 10 ms delay
net.cell(0).insert(Na())
net.cell(0).insert(K())
net.cell(0).insert(Leak())
net.cell(1).insert(Leak())
net.cell(2).insert(Leak())
net.cell(range(3)).branch(0).loc(0.0).record()
current = jx.step_current(i_delay, i_dur, i_amp, dt, t_max=100)
net.cell(0).branch(0).loc(0.0).stimulate(current)
trace = jx.integrate(net, delta_t=dt)
plt.plot(jnp.arange(trace.shape[1])*dt, trace.T)
plt.show()
```

![Image](https://github.com/user-attachments/assets/8a3c3c57-4f32-4c1e-8139-314f50e0af09)

will be very slow though for bigger delays until #632  gets fixed
